### PR TITLE
Add new monitor diagnostic setting modules

### LINKED ISSUE #529: Add new module for diagnostic settings
##### SUMMARY

Add new module for diagnostic settings

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

azure_rm_diagnosticsetting

##### ADDITIONAL INFORMATION

Add a new module to make provisioning diagnostic settings for resources possible.

```yaml
  - name: Create Diagnostic Setting
    azure_rm_diagnosticsetting:
      name: myDiagnosticSetting
      resource_id: /subscriptions/1234abc0-1234-5678-90ab-cdefghijklmn/resourceGroups/myResourceGroup/providers/Microsoft.Network/applicationGateways/myApplicationGateway
      storage_account_id: /subscriptions/1234abc0-1234-5678-90ab-cdefghijklmn/resourceGroups/myResourceGroup/providers/Microsoft.Storage/storageAccounts/myStorageAccount
      logs: 
        - category: ApplicationGatewayFirewallLog
          retention_policy:
            days: 0
```

### PR DESCRIPTION
**Pull request by l3ender**
##### SUMMARY
This PR adds support for configuring diagnostic settings for any Azure resource. Log and/or metric information from the resource can be sent to a variety of destinations.

An example which sends log information from an app service to an event hub:
```yaml
- name: Create diagnostic setting for webapp to an event hub
  azure_rm_monitordiagnosticsetting:
    name: "webapp-logs"
    resource:
      name: "my-webapp"
      type: "Microsoft.Web/sites"
      resource_group: "my-webapp-resource-group"
    event_hub:
      namespace: "my-event-hub"
      policy: "RootManageSharedAccessKey"
    logs:
      - category: "AppServiceConsoleLogs"
      - category: "AppServiceAppLogs"
```

Fixes https://github.com/ansible-collections/azure/issues/529.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_monitordiagnosticsetting
azure_rm_monitordiagnosticsetting_info

##### ADDITIONAL INFORMATION
I have added extensive test coverage, which can be run using a playbook like the following:

```yaml
---
- name: "Playbook for testing."
  hosts: "localhost"
  connection: "local"
  gather_facts: false
  vars:
    resource_group: "automated-testing"
    resource_group_secondary: "automated-testing-secondary"
  collections:
    - azure.azcollection

  tasks:
    - name: "Include tests"
      include_tasks: "tests/integration/targets/azure_rm_monitordiagnosticsetting/tasks/main.yml"
```

### CODE REVIEW

**Comment by l3ender** [comment]
Hi @Fred-sun, here is a module which I think will be very helpful for people. I saw https://github.com/ansible-collections/azure/pull/530 addressed the [same issue](https://github.com/ansible-collections/azure/issues/529), but the PR was quite stale and did not offer much flexibility. So I have raised this PR and hope it will be useful.

Please review and advise--thank you!

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+            retention_policy:
+                description:
+                    - The retention policy for this log.
+                type: dict
+                suboptions:
+                    days:
+                        description:
+                            - The number of days for the retention policy.
+                        type: int
+                    enabled:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py:109]: retention_policy defines default as (True) in argument, But this not!

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+        suboptions:
+            category:
+                description:
+                    - Name of a Management Group Diagnostic Log category for a resource type this setting is applied to.
+                type: str
+            category_group:
+                description:
+                    - Name of a Management Group Diagnostic Log category group for a resource type this setting is applied to.
+                type: str
+            enabled:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py:94]: enabled defines default as (True) in argument, But this not!

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+            retention_policy:
+                description:
+                    - The retention policy for this metric.
+                type: dict
+                suboptions:
+                    days:
+                        description:
+                            - The number of days for the retention policy.
+                        type: int
+                    enabled:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py:140]: enabled defines default as (True) in argument, But this not!

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+            - The list of metric setttings.
+            - At least one of I(metrics) or I(logs) must be specified for the diagnostic setting.
+        type: list
+        elements: dict
+        suboptions:
+            category:
+                description:
+                    - Name of a Diagnostic Metric category for a resource type this setting is applied to.
+                type: str
+            enabled:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py:125]: enabled defines default as (True) in argument, But this not!

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+                description:
+                    - The event hub namespace.
+                type: str
+                required: true
+            policy:
+                description:
+                    - The shared access policy.
+                type: str
+                required: true
+            namespace:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py]: The namespace is repeated with line 56, and the hub is missing from the event_hub suboption, which is defined in the arg

**Review by Fred-sun — COMMENTED** [review]
@l3ender A lot of normative questions about module documentation, Please refer to the link - https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#developing-modules-documen ting

File: plugins/modules/azure_rm_monitordiagnosticsetting_info.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+            returned: always
+            type: str
+            sample: my-diagnostic-setting
+        logs:
+            description:
+                - Enabled log configurations for the diagnostic setting.
+            returned: always
+            type: list
+            elements: dict
+            suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting_info.py]:
```suggestion
            contains:
```

File: plugins/modules/azure_rm_monitordiagnosticsetting_info.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+                    description:
+                        - Whether this log is enabled.
+                    type: bool
+                    returned: always
+                retention_policy:
+                    description:
+                        - The retention policy for this log.
+                    type: dict
+                    returned: always
+                    suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting_info.py]:
```suggestion
                    contains:
```

File: plugins/modules/azure_rm_monitordiagnosticsetting_info.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+                                - The number of days for the retention policy.
+                            type: int
+                            returned: always
+        metrics:
+            description:
+                - Enabled metric configurations for the diagnostic setting.
+            returned: always
+            type: list
+            elements: dict
+            suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting_info.py]:
```suggestion
            contains:
```

File: plugins/modules/azure_rm_monitordiagnosticsetting_info.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+                    description:
+                        - Whether the metric category is enabled.
+                    type: bool
+                    returned: always
+                retention_policy:
+                    description:
+                        - The retention policy for the metric category.
+                    type: dict
+                    returned: always
+                    suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting_info.py]:
```suggestion
                    contains:
```

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+            returned: always
+            type: str
+            sample: my-diagnostic-setting
+        logs:
+            description:
+                - Enabled log configurations for the diagnostic setting.
+            returned: always
+            type: list
+            elements: dict
+            suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py]:
```suggestion
            contains:
```

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+                    description:
+                        - Whether this log is enabled.
+                    type: bool
+                    returned: always
+                retention_policy:
+                    description:
+                        - The retention policy for this log.
+                    type: dict
+                    returned: always
+                    suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py]:
```suggestion
                    contains:
```

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 12]
```diff
@@ -0,0 +1,634 @@
...
+                    description:
+                        - Whether the metric category is enabled.
+                    type: bool
+                    returned: always
+                retention_policy:
+                    description:
+                        - The retention policy for the metric category.
+                    type: dict
+                    returned: always
+                    suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py]:
```suggestion
                    contains:
```

File: plugins/modules/azure_rm_monitordiagnosticsetting.py [hunk 13]
```diff
@@ -0,0 +1,634 @@
...
+                                - The number of days for the retention policy.
+                            type: int
+                            returned: always
+        metrics:
+            description:
+                - Enabled metric configurations for the diagnostic setting.
+            returned: always
+            type: list
+            elements: dict
+            suboptions:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_monitordiagnosticsetting.py]:
```suggestion
            contains:
```

**Review by Fred-sun — COMMENTED** [review]
@l3ender The return value suboption paramter is "contain", not suboptions. Thank you!

**Comment by l3ender** [comment]
Thanks for the review, @Fred-sun, and sorry for the silly documentation syntax mistakes. Thanks for providing the URL for detail on how to generate documentation; I will be looking more into that.

Please take another look. Thank you!

**Comment by l3ender** [comment]
Hello, wondering if there is anything else I can do for this PR? Thank you!

**Comment by Fred-sun** [comment]
> Hello, wondering if there is anything else I can do for this PR? Thank you!

@l3ender We are reviewing and will push forward the merger as soon as possible. Thank you very much!

**Comment by l3ender** [comment]
Hello, is anything else desired for this PR? Thank you!

**Comment by xuzhang3** [comment]
@l3ender  LGTM, shipping it.

**Pull request merged**